### ISSUE TITLE
In v. 2.9 `map_dom` always returns the 'root domain' even if a subdomain matches

### ISSUE THREAD

**Issue by bvli** [haproxy/haproxy#2400]
### Detailed Description of the Problem

Given the file `hosts.map`:
```
sub.example.com  be2
example.com      be1
```

When looking up `sub.example.com` (or just anything that contains `example.com`) using `map_dom(...)` "be1" is always returned.

### Expected Behavior

"be2" should be returned. It also does in versions < 2.9.

### Steps to Reproduce the Behavior

Using the `hosts.map` given above: 
```
sub.example.com  be2
example.com      be1
```
and the config below:

```
global
  stats socket /var/run/api.sock user haproxy group haproxy mode 660 level admin expose-fd listeners
frontend http
  bind *:8008
  mode http
  use_backend %[req.hdr(host),lower,map_dom(/usr/local/etc/haproxy/hosts.map,)]

backend be1
  mode http
  http-request return status 200 hdr x-backend be1

backend be2
  mode http
  http-request return status 200 hdr x-backend be2
```

Running haproxy 2.9 in docker like this:

```sh
$ docker run -d --rm --name ha -v .:/usr/local/etc/haproxy:ro -v /var/run/haproxy:/var/run -p 8008:8008 haproxytech/haproxy-alpine:2.9

$ echo "get map /usr/local/etc/haproxy/hosts.map sub.example.com" | socat stdio /var/run/haproxy/api.sock
```
returns:
```
type=dom, case=sensitive, found=yes, idx=list, key="example.com", value="be1", type="str"
```
While running the same commands but using haproxy 2.8 like this:

```sh
[ $ docker stop ha ]

$ docker run -d --rm --name ha -v .:/usr/local/etc/haproxy:ro -v /var/run/haproxy:/var/run -p 8008:8008 haproxytech/haproxy-alpine:2.8

$ echo "get map /usr/local/etc/haproxy/hosts.map sub.example.com" | socat stdio /var/run/haproxy/api.sock
```

returns: 
```
type=dom, case=sensitive, found=yes, idx=list, key="sub.example.com", value="be2", type="str"
```

When an actual request is made it is still the same behavior:
```sh
$ docker run -d --rm --name ha -v .:/usr/local/etc/haproxy:ro -v /var/run/haproxy:/var/run -p 8008:8008 haproxytech/haproxy-alpine:2.9

$ curl -i -H "Host: sub.example.com" http://localhost:8008
HTTP/1.1 200 OK
x-backend: be1
content-length: 0

```

Vs: 

```
docker run -d --rm --name ha -v .:/usr/local/etc/haproxy:ro -v /var/run/haproxy:/var/run -p 8008:8008 haproxytech/haproxy-alpine:2.8

curl -i -H "Host: sub.example.com" http://localhost:8008
HTTP/1.1 200 OK
x-backend: be2
content-length: 0
```

### Do you have an idea how to solve the issue?

The workaround right now is to use `map(..)` instead of `map_dom(..)` and then miss the subdomain wildcard functionality. Personally I've downgraded to 2.8.

### What is your configuration?

```haproxy
global
  stats socket /var/run/api.sock user haproxy group haproxy mode 660 level admin expose-fd listeners
frontend http
  bind *:8008
  mode http
  use_backend %[req.hdr(host),lower,map_dom(/usr/local/etc/haproxy/hosts.map,)]

backend be1
  mode http
  http-request return status 200 hdr x-backend be1

backend be2
  mode http
  http-request return status 200 hdr x-backend be2
```

### Output of `haproxy -vv`

```plain
[$ docker exec ha haproxy -vv]
HAProxy version 2.9.1-f72603c 2023/12/15 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.9.1.html
Running on: Linux 5.15.133.1-microsoft-standard-WSL2 #1 SMP Thu Oct 5 21:02:42 UTC 2023 x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PTHREAD_EMULATION=1 USE_LINUX_TPROXY=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_TFO=1 USE_QUIC=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_QUIC_OPENSSL_COMPAT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 -BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX +PTHREAD_EMULATION +QUIC +QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=8).
Built with OpenSSL version : OpenSSL 3.1.4 24 Oct 2023
Running on OpenSSL version : OpenSSL 3.1.4 24 Oct 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.1 20220924

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

**Comment by Darlelet** [comment]
Hi,

Thanks for the report!

Based on your example, a quick bisect shows that the change of behavior was introduced during 2.9 development by commit 5fea59754b7765ca83aea3813f8b375160c7f3b0

Indeed, using the same input map file, the order of elements loaded by `pattern_read_from_file()` seem to differ before and after the above commit. Using your example map file, prior to the faulty commit, "sub.example.com" is loaded first while after the faulty commit "example.com" is loaded first.

**Comment by Darlelet** [comment]
It seems like this is a side effect due to using eb tree for the lookup:

Prior to https://github.com/haproxy/haproxy/commit/5fea59754b7765ca83aea3813f8b375160c7f3b0, the ordering depended on the initial position of the element in the map file (the order in which elements are seen), whereas the ordering now strictly relies on `ebis_insert()`

**Comment by Darlelet** [comment]
It looks like all list-based matching methods are affected, `dom` is one of them.

tree-based matching methods (ie: `str`, `beg`, `ip`) are not affected since the initial ordering was already lost when the node is inserted.

I can't find any comment in the code or the documentation that tells whether or not initial map file ordering should be respected for list-based matching types. But if we wan't to stay conservative with existing behavior, we may want to partially revert https://github.com/haproxy/haproxy/commit/5fea59754b7765ca83aea3813f8b375160c7f3b0 so that `pattern_read_from_file()` uses the list to iterate on elements instead of the tree. Since we loop on ALL elements, there was no performance benefit from using the tree to iterate over elements anyway.

The patch below should help:
[0001-BUG-MINOR-map-list-based-matching-potential-ordering.patch.txt](https://github.com/haproxy/haproxy/files/13819124/0001-BUG-MINOR-map-list-based-matching-potential-ordering.patch.txt)

**Comment by bvli** [comment]
Just noting that I had had the impression that the order of entry in the list - when using  `map_com(..)`  - was relevant (and should be respected). I was certain I had seen it somewhere in the documentation.  Here I'm referring to this sentence in the patch:

> So while it may not be considered as a real bug since neither the code or the documentation officially states that the file ordering should be respected

Even if a blog post is not official documentaiton, I'd say that the [blog post](https://www.haproxy.com/blog/how-to-map-domain-names-to-backend-server-pools-with-haproxy) explaining how to use maps to map backends would loose relevance, if the above statement is true, as the first map shown in the [HA Proxy Maps](https://www.haproxy.com/blog/how-to-map-domain-names-to-backend-server-pools-with-haproxy#haproxy-maps) section, would always return `be_default` even when the requested host is `api.example.net|com`, `api[1..10000].example.net|com` . I don't think this is the intent and it is certainly not what is implied in the blog post 🙂 

Thanks for fixing the bug! I'm looking forward to the backport 👍

**Comment by Darlelet** [comment]
Thanks for pointing it out. After a quick talk with other haproxy developers it was confirmed that mapfile ordering should be preserved for list-based matching types such as `dom` and that we need to consider this regression as an actual bug. Thus, the above fix will be backported to 2.9

**Comment by Darlelet** [comment]
Just a quick update to inform you that 2.9.2 was released with the fix :)
I guess we can close the issue
